### PR TITLE
Reduce allocations in FeatureFlags.IsEnabled

### LINKED ISSUE #10883: Optimize FeatureFlags.IsEnabled
The `FeatureFlags.IsEnabled` method currently uses `string.Split`, which results in a string array allocation. Instead, we can optimize this by using `ReadOnlySpan<char>` to check if the requested feature name is present, avoiding the extra allocation.

https://github.com/Azure/azure-functions-host/blob/5a873034be67cb2300f008bb39ef75fbc9e8b63e/src/WebJobs.Script/Config/FeatureFlags.cs#L22-L23

### PR DESCRIPTION
**Pull request by kshyju**
Fixes #10883

Optimizes memory allocations in `FeatureFlags.IsEnabled` by adopting the zero-allocation `ContainsToken` extension method for efficient token search in delimited strings.

This method gets called many times during specialization.

Allocation Snapshot

I tested with a non-empty value for `AzureWebJobsFeatureFlags`.

```json
"AzureWebJobsFeatureFlags": "EnableWorkerIndexing"
```

#### Before

| Name                  | Total Allocations | Self Allocations | Total Size (Bytes) | Self Size (Bytes) |
|-----------------------|-------------------|------------------|--------------------|-------------------|
| `FeatureFlags.IsEnabled` | 645               | 0                | 27,090             | 0                 |

#### After

| Name                  | Total Allocations | Self Allocations | Total Size (Bytes) | Self Size (Bytes) |
|-----------------------|-------------------|------------------|--------------------|-------------------|
| `FeatureFlags.IsEnabled` | 215               | 0                | 13,330             | 0                 |

#### Delta
- `FeatureFlags.IsEnabled(string, IEnvironment)` – Total allocations dropped by **430 (66.67%)**
- `FeatureFlags.IsEnabled(string, IEnvironment)` – Total allocated size dropped by **13,760 bytes (50.80%)**

---

### Benchmark.NET Micro Benchmarks

Also ran micro benchmark tests to compare the performance of the new implementation with the old one.
source code used: https://github.com/kshyju/Benchmarks/blob/9fe638a43458591991317a72af4b7b7cd30ed6a9/src/Benchmarks.ConsoleApp/StringSplitBenchmarks.cs#L6

#### Windows  
[Benchmark Report](https://github.com/kshyju/Benchmarks/actions/runs/15189525120/job/42718386961#step:7:1395)

| Method                                 | Mean      | Allocated |
|----------------------------------------|-----------|-----------|
| `ContainsUsingStringSplit_Short_Present`   | 41.45 ns | 64 B      |
| `ContainsToken_Short_Present`             | 11.52 ns | -         |
| `ContainsUsingStringSplit_Short_Absent`    | 44.09 ns | 64 B      |
| `ContainsToken_Short_Absent`              | 10.59 ns | -         |
| `ContainsUsingStringSplit_Medium_Present`  | 121.10 ns | 272 B     |
| `ContainsToken_Medium_Present`            | 31.06 ns | -         |
| `ContainsUsingStringSplit_Medium_Absent`   | 118.37 ns | 280 B     |
| `ContainsToken_Medium_Absent`             | 32.21 ns | -         |
| `ContainsUsingStringSplit_Empty`          | 0.33 ns  | -         |
| `ContainsToken_Empty`                     | 1.59 ns  | -         |

#### Linux (Ubuntu)  
[Benchmark Report](https://github.com/kshyju/Benchmarks/actions/runs/15189525120/job/42718386079#step:7:1384)

| Method                                 | Mean      | Allocated |
|----------------------------------------|-----------|-----------|
| `ContainsUsingStringSplit_Short_Present`   | 39.69 ns | 64 B      |
| `ContainsToken_Short_Present`             | 10.58 ns | -         |
| `ContainsUsingStringSplit_Short_Absent`    | 42.72 ns | 64 B      |
| `ContainsToken_Short_Absent`              | 10.63 ns | -         |
| `ContainsUsingStringSplit_Medium_Present`  | 135.73 ns | 272 B     |
| `ContainsToken_Medium_Present`            | 31.16 ns | -         |
| `ContainsUsingStringSplit_Medium_Absent`   | 138.68 ns | 280 B     |
| `ContainsToken_Medium_Absent`             | 32.59 ns | -         |
| `ContainsUsingStringSplit_Empty`          | 0.35 ns  | -         |
| `ContainsToken_Empty`                     | 1.26 ns  | -         |

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [ ] My changes **do not** require documentation changes
    * [x] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+using Microsoft.Azure.WebJobs.Script.Utils;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Script.Tests.Utils
+{
+    public sealed class StringUtilsTests
+    {
+        [Theory]
+        [InlineData("FeatureA,FeatureB,FeatureC", "FeatureB", ',', true)]
```
Comment (RohitRanjanMS) [inline test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs]: Should we account for whitespace and empty tokens? 

"FeatureA, FeatureB,FeatureC"

File: src/WebJobs.Script/Utils/StringUtils.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            var searchSpan = searchToken.AsSpan();
+
+            while (!remaining.IsEmpty)
+            {
+                var separatorIndex = remaining.IndexOf(separator);
+                ReadOnlySpan<char> currentToken;
+
+                if (separatorIndex >= 0)
+                {
+                    currentToken = remaining.Slice(0, separatorIndex);
```
Comment (RohitRanjanMS) [inline src/WebJobs.Script/Utils/StringUtils.cs]: Comparing currentToken and searchToken here should save some nanoseconds. You can skip a slice if currentToken matches the target.

**Review by RohitRanjanMS — CHANGES_REQUESTED** [review]

File: test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs (hunk 1 above)
Comment (kshyju) [reply test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs]: I intentionally left that out to maintain parity with the previous version, which also didn’t handle trimming. In my benchmark repository, I’ve included a unit test that compares the outputs of both implementations to ensure the functionality remains consistent.

https://github.com/kshyju/Benchmarks/blob/9fe638a43458591991317a72af4b7b7cd30ed6a9/tests/Benchmarks.Tests/Tests.cs#L17-L26

File: src/WebJobs.Script/Utils/StringUtils.cs (hunk 2 above)
Comment (kshyju) [reply src/WebJobs.Script/Utils/StringUtils.cs]: Okay, I tried [that approach](https://github.com/kshyju/Benchmarks/blob/471002fec0a589bd1a1b81ff395e22c034968353/src/Benchmarks.ConsoleApp/StringUtils.cs#L66-L69) and ran benchmarks to compare those 2.

##### Linux 
https://github.com/kshyju/Benchmarks/actions/runs/15201876152/job/42757382146#step:7:1378

| Method                                   | Mean      |
|-----------------------------------------|-----------|
| ContainsToken_Short_Present              | 11.548 ns |
| ContainsToken_CheckInline_Short_Present  | 10.645 ns |
| ContainsToken_Short_Absent               | 10.644 ns |
| ContainsToken_CheckInline_Short_Absent   | 10.330 ns |
| ContainsToken_Medium_Present             | 31.076 ns |
| ContainsToken_CheckInline_Medium_Present | 31.335 ns |
| ContainsToken_Medium_Absent              | 32.303 ns |
| ContainsToken_CheckInline_Medium_Absent  | 32.730 ns |
| ContainsToken_Empty                      |  1.332 ns |
| ContainsToken_CheckInline_Empty          |  1.338 ns |


##### Windows
https://github.com/kshyju/Benchmarks/actions/runs/15201876152/job/42757382306#step:7:1387

| Method                                   | Mean      |
|-----------------------------------------|-----------|
| ContainsToken_Short_Present              | 11.196 ns |
| ContainsToken_CheckInline_Short_Present  | 10.909 ns |
| ContainsToken_Short_Absent               | 10.588 ns |
| ContainsToken_CheckInline_Short_Absent   | 10.291 ns |
| ContainsToken_Medium_Present             | 30.436 ns |
| ContainsToken_CheckInline_Medium_Present | 36.290 ns |
| ContainsToken_Medium_Absent              | 31.417 ns |
| ContainsToken_CheckInline_Medium_Absent  | 30.823 ns |
| ContainsToken_Empty                      |  1.586 ns |
| ContainsToken_CheckInline_Empty          |  1.593 ns |

Comparing them:

| Platform | Input   | Match   | Variable (ns) | Inline (ns) | Δ (ns) | Winner  |
|----------|---------|---------|---------------|-------------|--------|---------|
| Linux    | Short   | Present | 11.55         | 10.65       | 0.90   | Inline  |
| Linux    | Short   | Absent  | 10.64         | 10.33       | 0.31   | Inline  |
| Linux    | Medium  | Present | 31.08         | 31.34       | -0.26  | Variable |
| Linux    | Medium  | Absent  | 32.30         | 32.73       | -0.43  | Variable |
| Linux    | Empty   | —       | 1.33          | 1.34        | -0.01  | Tie     |
| Windows  | Short   | Present | 11.20         | 10.91       | 0.29   | Inline  |
| Windows  | Short   | Absent  | 10.59         | 10.29       | 0.30   | Inline  |
| Windows  | Medium  | Present | 30.44         | 36.29       | -5.85  | Variable |
| Windows  | Medium  | Absent  | 31.42         | 30.82       | 0.60   | Inline  |
| Windows  | Empty   | —       | 1.59          | 1.59        | 0.00   | Tie     |

TLDR:
- The inline version is slightly faster for short inputs, but the difference is typically less than 1 nanosecond.  
- For medium and empty inputs, performance is nearly identical, with the variable version sometimes ahead.

I personally prefer the variable version (currently in the PR) since I find it more readable with 2 sections. The top section snippet handles token extraction, and the bottom handles the comparison. Do you find the inline version more readable? I am open to switching to that if more people find that readable.

Comment (RohitRanjanMS) [reply src/WebJobs.Script/Utils/StringUtils.cs]: Interesting, I didn't expect inline version to take longer compared to the variable version especially when there's a match (we will save one slice). I find the variable version more readable as you would need another level of nesting in the inline version. I am fine if you want to keep your existing version.

Comment (RohitRanjanMS) [reply src/WebJobs.Script/Utils/StringUtils.cs]: I ran your test locally, inline is promising in all the scenarios.

image

**Review by safihamid — APPROVED** [review]
:shipit:

File: src/WebJobs.Script/Utils/StringUtils.cs (hunk 2 above)
Comment (kshyju) [reply src/WebJobs.Script/Utils/StringUtils.cs]: Yea, I expect results from local dev workstation to have higher variance/noise (std dev) than the one on servers. From all these 3 results, looks like the Linux server has the least variance. I still vote for readability over 5 ns.

**Review by RohitRanjanMS — APPROVED** [review]

**Comment by jviau** [comment]
.NET8 has a type called `SearchValues<T>` which I think can be used for highly optimized string searching. Maybe this can help here? Not entirely sure how to use it or if it is applicable, but worth a look.

**Comment by kshyju** [comment]
> .NET8 has a type called `SearchValues<T>` which I think can be used for highly optimized string searching. Maybe this can help here? Not entirely sure how to use it or if it is applicable, but worth a look.

I think `SearchValues<T>` is useful when we have to search using multiple delimiters. It still causes some allocation with the use of ` SearchValues.Create`. In the context of this method, it will not offer anything better since we care about single char delimiter. I ran some benchmarks to compare both approaches, and the results confirm that using SearchValues<T> here does not do better than the char approach we have in the `ContainsToken` method.

1. [Linux](https://github.com/kshyju/Benchmarks/actions/runs/15283035314/job/42986465395#step:7:1384)
2. [Windows](https://github.com/kshyju/Benchmarks/actions/runs/15283035314/job/42986465379#step:7:1386)

File: test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        [InlineData("FeatureA,FeatureB,FeatureC", "FeatureD", ',', false)]
+        [InlineData("FeatureA,FeatureB,FeatureC", "featureb", ',', true)]
+        [InlineData("FeatureA|FeatureB|FeatureC", "FeatureC", '|', true)]
+        [InlineData("FeatureA,FeatureB,FeatureC", "FeatureA", ',', true)]
+        [InlineData("FeatureA,FeatureB,FeatureC", "FeatureC", ',', true)]
+        [InlineData("FeatureA,FeatureB,FeatureC", "FeatureA,FeatureB", ',', false)]
+        [InlineData(null, "FeatureA", ',', false)]
+        [InlineData("", "FeatureA", ',', false)]
+        public void ContainsToken_LargeInputScenarios(string delimited, string token, char separator, bool expected)
```
Comment (jviau) [inline test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs]: Some edge cases to also test:

- multiple separators: `FeatureA,,FeatureB`
- leading separator: `,FeatureA`
- trailing separator: `FeatureA,`
- only separator: `,`

File: src/WebJobs.Script/Utils/StringUtils.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        /// <summary>
+        /// Determines whether a delimited string contains a specific token, using the specified separator and string comparison.
+        /// This method is a zero-allocation, faster alternative to splitting the string and using Contains, as it avoids unnecessary allocations.
+        /// </summary>
+        /// <param name="delimitedString">The string containing delimited tokens to search(Ex: "FeatureA,FeatureB").</param>
+        /// <param name="searchToken">The token to search for within the delimited string.</param>
+        /// <param name="separator">The character used to separate tokens in the string. Defaults to ','.</param>
+        /// <param name="comparisonType">The string comparison type to use. Defaults to OrdinalIgnoreCase.</param>
+        /// <returns>True if the token is found; otherwise, false.</returns>
+        internal static bool ContainsToken(string delimitedString, string searchToken, char separator = ',', StringComparison comparisonType = StringComparison.OrdinalIgnoreCase)
```
Comment (jviau) [inline src/WebJobs.Script/Utils/StringUtils.cs]: Is it worth taking in `ReadOnlySpan<char>` for the arguments?

```suggestion
        internal static bool ContainsToken(ReadOnlySpan<char> source, ReadOnlySpan<char> token, char separator = ',', StringComparison comparisonType = StringComparison.OrdinalIgnoreCase)
```

Also changed some param names for conciseness.

Maybe even make this an extension method for `string`?

**Review by jviau — APPROVED** [review]

File: test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs (hunk 3 above)
Comment (kshyju) [reply test/WebJobs.Script.Tests/Utils/StringUtilsTests.cs]: Improved test coverage.

File: src/WebJobs.Script/Utils/StringUtils.cs (hunk 4 above)
Comment (kshyju) [reply src/WebJobs.Script/Utils/StringUtils.cs]: Changed param names to concise ones.

Comment (kshyju) [reply src/WebJobs.Script/Utils/StringUtils.cs]: Switched to extension method. Made another overload which takes ReadOnlySpan<char> as well.

File: src/WebJobs.Script/Extensions/TokenExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        /// <param name="token">The token to search for. Must not contain the separator character. A match is determined using the specified comparison type.</param>
+        /// <param name="separator">The character used to separate tokens in the string. Defaults to ','.</param>
+        /// <param name="comparisonType">The string comparison type to use. Defaults to OrdinalIgnoreCase.</param>
+        /// <returns>
+        /// <c>true</c> if the token is found as an exact match in the delimited string; otherwise, <c>false</c>.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown if <paramref name="token"/> contains the separator character.
+        /// </exception>
+        public static bool ContainsToken(this string source, string token, char separator = ',', StringComparison comparisonType = StringComparison.OrdinalIgnoreCase)
```
Comment (jviau) [inline src/WebJobs.Script/Extensions/TokenExtensions.cs]: Final nit: consider removing the default param on `separator`.

My reasoning is the method name "ContainsToken" (which I like, no need to change it), doesn't define what a "token" is exactly. By having the caller always supply `separator`, this forces them to acknowledge a separator is in play here. And hopefully they understand a _token_ are the parts between the separators.

**Review by jviau — APPROVED** [review]

File: src/WebJobs.Script/Extensions/TokenExtensions.cs (hunk 5 above)
Comment (kshyju) [reply src/WebJobs.Script/Extensions/TokenExtensions.cs]: I agree with your reasoning. 
Pushed an update to remove default value for `separator` parameter.

**Review by jviau — APPROVED** [review]
